src/utils/utilities.jl

```diff
@@ -63,3 +63,30 @@ function meshgrid(vx::AbstractVector{T}, vy::AbstractVector{T},
 	oo = ones(Int, o)
 	(vx[om, :, oo], vy[:, on, oo], vz[om, on, :])
 end
+
+"""
+    mean(f, A, region)
+
+Apply the function `f` to each element of `A`, and compute the mean along dimension in `region`.
+"""
+function Base.mean(f::Function, a::AbstractArray, region::Int)
+    x = Base.mapreducedim(f, +, a, region)
+    n = max(1, Base._length(x)) // Base._length(a)
+    x .= x .* n
+
+    return x
+end
+
+"""
+    mean!(f, r, A)
+
+Apply `f` to each element of A, and compute the mean over the singleton dimensions of `r`, and write the results to `r`.
+"""
+function Base.mean!(f::Function, r::AbstractArray{T}, a::AbstractArray) where {T<:Number}
```

Comment:
is there a reason why the element type of a is not fixed?

Reply:
I did that to avoid tracking the domain and range of `f`, so that `f` can output a different type than `a`. 

An example would be `f` taking a `Float32` from `a` as input, and outputting a `Complex{Float32}` into the pre-allocated `r`